ayeletK/assignment2/Q2_solution.cpp

```diff
@@ -0,0 +1,71 @@
+#include "tree_node.cpp"
+#include <string>
+
+using namespace std;
+
+#define KEY_NOT_IN_PATH "-"
+#define EMPTY_PATH ""
+#define LEFT_SON '0'
+#define RIGHT_SON '1'
+
+template <typename Key>
+char get_addition_to_path(Node<Key>* son_node, Node<Key>* parent_node){
+  return parent_node->left == son_node ? LEFT_SON : RIGHT_SON;
+}
+
+template <typename Key>
+string get_path_string(Key key, Node<Key>* curr_node, Node<Key>* parent_node = NULL){
+  if (!curr_node) return KEY_NOT_IN_PATH;
+  Key curr_key = curr_node->get_key();
+  // EMPTY_PATH represents the case the root node is the key
+  if (curr_key == key){
+    if (!parent_node) return EMPTY_PATH;
+    else return string(1, get_addition_to_path(curr_node, parent_node));
+  }
+  string res;
+  res = get_path_string(key, curr_node->left, curr_node);
+  if (res == KEY_NOT_IN_PATH){
+    res = get_path_string(key, curr_node->right, curr_node);
+  }
+  if (res != KEY_NOT_IN_PATH && parent_node){
+    char addition = get_addition_to_path(curr_node, parent_node);
+    res.insert(res.begin(), addition);
+  }
+  return res;
+}
+
+template <typename Key>
+Node<Key>* get_node_from_path(Node<Key>* root, string path){
+  if (!root) throw "the given tree is empty.";
+  if (path == KEY_NOT_IN_PATH) return NULL;
+  if (path == EMPTY_PATH) return root;
```

Comment:
You can omit this line, `root` will still be returned.

Reply:
Right! thanks :)